### ISSUE TITLE
ergonomic handling of scalars (singletons)

### ISSUE THREAD

**Issue by jhellerstein** [hydro-project/hydroflow#797]
We currently assume every Hydroflow expression is a collection -- specifically an ordered stream.

This assumption makes it very annoying to write simple scalar expressions. For example, in Paxos we have boolean expressions with patterns like the ones in the large code block below. We should aim to (a) make these ergonomic, and (b) encourage/identify monotonicity where possible in these expressions.

Related goal: allow scalar Hydroflow values to be referenced in streaming expressions. For example:
```
the_sum = source_iter(vec![1,2,3]) -> reduce(|accum, elem| { accum += elem; accum }); // a singleton
stream = source_iter(vec![5,6,7]); // a stream
stream -> filter(|x| x > the_sum) -> inspect();
```
Currently we have to do the above via a cross-join that tuples up the singleton element in `the_sum` with each element in the `stream`.

---

Here are the exercises from paxos:
```
use std::cmp::Ordering;
use std::iter::Iterator;

use clap::{Parser, ValueEnum};
use hydroflow::scheduled::graph::Hydroflow;
use hydroflow::{hydroflow_syntax, tokio};
use lattices::{IsTop, Max};

#[derive(Parser, Debug, Clone, ValueEnum)]
enum GraphType {
    Mermaid,
    Dot,
    Json,
}

#[derive(Parser, Debug, Clone)]
struct Opts {
    #[clap(value_enum, long)]
    graph: Option<GraphType>,
}

pub(crate) fn two_ands() -> Hydroflow {
    hydroflow_syntax! {
        // if (X and Y and Z) then do A
        X = source_iter(vec![Max::<bool>::new(true)]);
        Y = source_iter(vec![Max::<bool>::new(true)]);
        Z = source_iter(vec![Max::<bool>::new(true)]);
        A = source_iter(vec!["do A"]) -> [0]doit;

        X -> [0]and_1;
        Y -> [1]and_1;
        and_1 -> [0]and_2;
        Z -> [1]and_2;
        and_1 = cross_join::<'tick>() -> filter(|(x, _)| x.is_top()) -> filter(|(_, y)| y.is_top()) -> map(|(x, _)| x);
        and_2 = cross_join::<'tick>() -> filter(|(x, _)| x.is_top()) -> filter(|(_, y)| y.is_top()) -> map(|(x, _)| x);
        and_2
          -> assert([Max::<bool>::new(true)])
          -> [1]doit;
        doit = cross_join() -> map(|(x, _)| x)
          -> for_each(|x| println!("result: {:?}", x));
    }
}

pub(crate) fn greater_than() -> Hydroflow {
    hydroflow_syntax! {
        // if (X > 8) then do B
        X = source_iter(vec![Max::new(9)]);
        A = source_iter(vec!["do B"]) -> [0]doit;

        // in principle Lattice::cmp(constant) == Ordering::Greater is monotonic,
        // but it's hard to check. Might be nicer to have a monotone function called `threshold`
        X -> filter(|x| x.cmp(&Max::new(8)) == Ordering::Greater)
          -> assert([Max::new(9)])
          -> [1]doit;

        doit = cross_join() -> map(|(x, _)| x)
          -> for_each(|x| println!("result: {:?}", x));
    }
}

pub(crate) fn x_and_empty_y() -> Hydroflow {
    hydroflow_syntax! {
        // if (X and Y.empty()) then do C
        X = source_iter(vec![Max::new(true)]);
        Y = source_iter(vec![]);
        A = source_iter(vec!["do C"]) -> [0]doit;

        X -> [0]and;
        // the next line is is_empty(). It's non-monotone in at least 2 ways (and it should be).
        // probably deserves a syntax sugar operator
        Y -> fold::<'tick>(Max::new(true), |_, _: Max::<bool>| Max::new(false))
          -> [1]and;
        and = cross_join::<'tick>() -> filter(|(x, _)| x.is_top()) -> filter(|(_, y)| y.is_top()) -> map(|(x, _)| x)
          -> assert([Max::new(true)])
          -> [1]doit;

        doit = cross_join() -> map(|(x, _)| x)
          -> for_each(|x| println!("result: {:?}", x));
    }
}

pub(crate) async fn run_flow(mut flow: Hydroflow, opts: Opts) {
    flow.run_available();

    if let Some(graph) = opts.graph {
        let serde_graph = flow
            .meta_graph()
            .expect("No graph found, maybe failed to parse.");
        match graph {
            GraphType::Mermaid => {
                println!("{}", serde_graph.to_mermaid());
            }
            GraphType::Dot => {
                println!("{}", serde_graph.to_dot())
            }
            GraphType::Json => {
                unimplemented!();
                // println!("{}", serde_graph.to_json())
            }
        }
    }
}

#[hydroflow::main]
async fn main() {
    // parse command line arguments
    let opts = Opts::parse();
    run_flow(two_ands(), opts.clone()).await;
    run_flow(greater_than(), opts.clone()).await;
    run_flow(x_and_empty_y(), opts.clone()).await;
}
```

**Comment by MingweiSamuel** [comment]
Come back to this as a quick 0.6.1

**Comment by MingweiSamuel** [comment]
How to handle join?
```rust
hydroflow_parser! {
    lhs = source_iter([(7, 1), (7, 2)]) -> state();
    rhs = source_iter([(7, 0), (7, 1), (7, 2)]) -> state();
    state_join(lhs, rhs)
        -> for_each(|x| println!("OUT: {:?}", x));
}

hydroflow_parser! {
    source_iter([(7, 1), (7, 2)]) -> state() -> [0]my_join;
    source_iter([(7, 0), (7, 1), (7, 2)]) -> state() -> [0]my_join;
    my_join = state_join(lhs, rhs)
        -> for_each(|x| println!("OUT: {:?}", x));
}
```

**Comment by MingweiSamuel** [comment]
#1065

**Comment by MingweiSamuel** [comment]
Implemented #1065 in hydroflow 0.6.1
